File: frontend/shader/contract/ts/aptos.ts
// ============================================================================
// Aptos Configuration
// ============================================================================

export interface AptosConfig {
  rpcUrl: string;
  network: "mainnet" | "testnet";
}

// ============================================================================
// Aptos API Response Types
// ============================================================================

export interface AptosAccountResponse {
  data?: {
    coin?: {
      value?: string;
    };
  };
}

export interface AptosTransactionResponse {
  hash?: string;
  transaction?: {
    hash?: string;
  };
}

/**
 * Aptos Connector
 * Connects to Aptos blockchain
 */
export class AptosConnector {
  private readonly config: AptosConfig;

  constructor(config: AptosConfig) {
    this.config = config;
  }

  async getBalance(address: string): Promise<string> {
    if (!this.isValidAddress(address)) {
      throw new Error("Invalid Aptos address");
    }

    try {
      let apiUrl: string;
      if (this.config.network === "mainnet") {
        apiUrl = this.config.rpcUrl || "https://fullnode.mainnet.aptoslabs.com";
      } else {
        apiUrl = this.config.rpcUrl || "https://fullnode.testnet.aptoslabs.com";
      }

      // Aptos uses REST API, not JSON-RPC
      const response = await fetch(
        `${apiUrl}/v1/accounts/${address}/resource/0x1::coin::CoinStore<0x1::aptos_coin::AptosCoin>`,
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );

      if (!response.ok) {
        // If account doesn't exist or has no balance, return 0
        if (response.status === 404) {
          return "0";
        }
        throw new Error("Failed to fetch Aptos balance");
      }

      const data = (await response.json()) as AptosAccountResponse;

      // Return balance in Octas (smallest unit, 1 APT = 100,000,000 Octas)
      return data.data?.coin?.value || "0";
    } catch (error) {
      throw new Error(
        `Aptos getBalance failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  async sendTransaction(
    fromAddress: string,
    toAddress: string,
    amount: string
  ): Promise<string> {
    if (!this.isValidAddress(fromAddress) || !this.isValidAddress(toAddress)) {
      throw new Error("Invalid Aptos address");
    }

    const amountNum = Number.parseInt(amount, 10);
    if (Number.isNaN(amountNum) || amountNum <= 0) {
      throw new Error("Invalid amount");
    }

    try {
      // Aptos transactions require:
      // 1. Building transaction with proper Move function call
      // 2. Signing with private key
      // 3. Submitting to network
      // This is simplified - in production use aptos SDK

      let apiUrl: string;
      if (this.config.network === "mainnet") {
        apiUrl = this.config.rpcUrl || "https://fullnode.mainnet.aptoslabs.com";
      } else {
        apiUrl = this.config.rpcUrl || "https://fullnode.testnet.aptoslabs.com";
      }

      // Get account sequence number
      const accountResponse = await fetch(`${apiUrl}/v1/accounts/${fromAddress}`, {
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!accountResponse.ok) {
        throw new Error("Failed to get account information");
      }

      // Submit transaction
      // Note: In production, use aptos SDK to properly build and sign
      const response = await fetch(`${apiUrl}/v1/transactions`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          // Signed transaction would go here
          // This requires proper serialization and signing
          sender: fromAddress,
          sequence_number: "0", // Would be fetched from account
          max_gas_amount: "1000",
          gas_unit_price: "100",
          expiration_timestamp_secs: "0", // Would be calculated
          payload: {
            type: "entry_function_payload",
            function: "0x1::coin::transfer",
            type_arguments: ["0x1::aptos_coin::AptosCoin"],
            arguments: [toAddress, amountNum.toString()],
          },
          signature: {
            type: "ed25519_signature",
            public_key: "0x...",
            signature: "0x...",
          },
        }),
      });

      if (!response.ok) {
        throw new Error("Failed to submit Aptos transaction");
      }

      const data = (await response.json()) as AptosTransactionResponse & {
        message?: string;
      };

      if (data.message) {
        throw new Error(data.message);
      }

      return data.hash || data.transaction?.hash || "transaction_hash";
    } catch (error) {
      throw new Error(
        `Aptos sendTransaction failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  isValidAddress(address: string): boolean {
    // Aptos addresses are 0x followed by 64 hex characters
    const aptosAddressPattern = /^0x[a-fA-F0-9]{64}$/;
    return aptosAddressPattern.test(address);
  }

  /**
   * Bridge tokens from Aptos to another chain
   */
  async bridgeTokens(
    sourceAddress: string,
    destinationAddress: string,
    amount: string
  ): Promise<{ sourceTx: string; destinationTx?: string }> {
    const txHash = await this.sendTransaction(
      sourceAddress,
      destinationAddress,
      amount
    );
    return { sourceTx: txHash };
  }

  /**
   * Get chain name for bridge operations
   */
  getChainName(): string {
    return "Aptos";
  }

  /**
   * Get chain aliases for bridge operations
   */
  getChainAliases(): string[] {
    return ["aptos", "apt"];
  }
}

File: store/rust/rust/aptos.rs
//! Aptos connector

use crate::{http_get_json, http_post_json, AptosNetwork, BridgeError, BridgeResult};
use serde::{Deserialize, Serialize};

/// Aptos connector
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AptosConnector {
    rpc_url: String,
    #[allow(dead_code)]
    network: AptosNetwork,
}

impl AptosConnector {
    pub fn new(rpc_url: &str, network: AptosNetwork) -> BridgeResult<Self> {
        Ok(Self { rpc_url: rpc_url.to_string(), network })
    }

    pub fn get_balance(&self, address: &str) -> BridgeResult<u64> {
        if !self.is_valid_address(address) {
            return Err(BridgeError::InvalidAddress(address.to_string()));
        }

        // Aptos uses REST API, not JSON-RPC
        let url = format!(
            "{}/v1/accounts/{}/resource/0x1::coin::CoinStore<0x1::aptos_coin::AptosCoin>",
            self.rpc_url.trim_end_matches('/'),
            address
        );

        match http_get_json(&url) {
            Ok(result) => {
                // Return balance in Octas (smallest unit, 1 APT = 100,000,000 Octas)
                let balance = result
                    .get("data")
                    .and_then(|d| d.get("coin"))
                    .and_then(|c| c.get("value"))
                    .and_then(|v| v.as_str())
                    .and_then(|s| s.parse::<u64>().ok())
                    .unwrap_or(0);
                Ok(balance)
            },
            Err(BridgeError::NetworkError(_)) => {
                // If account doesn't exist or has no balance, return 0
                Ok(0)
            },
            Err(e) => Err(e),
        }
    }

    pub fn send_transaction(
        &self,
        from_address: &str,
        to_address: &str,
        amount: u64, // in Octas
    ) -> BridgeResult<String> {
        if !self.is_valid_address(from_address) || !self.is_valid_address(to_address) {
            return Err(BridgeError::InvalidAddress("Invalid Aptos address".to_string()));
        }
        if amount == 0 {
            return Err(BridgeError::InvalidAmount("Amount must be greater than 0".to_string()));
        }

        // Aptos transactions require:
        // 1. Building transaction with proper Move function call
        // 2. Signing with private key
        // 3. Submitting to network
        // This is simplified - in production use aptos SDK
        let url = format!("{}/v1/transactions", self.rpc_url.trim_end_matches('/'));
        let body = serde_json::json!({
            "sender": from_address,
            "sequence_number": "0", // Would be fetched from account
            "max_gas_amount": "1000",
            "gas_unit_price": "100",
            "expiration_timestamp_secs": "0", // Would be calculated
            "payload": {
                "type": "entry_function_payload",
                "function": "0x1::coin::transfer",
                "type_arguments": ["0x1::aptos_coin::AptosCoin"],
                "arguments": [to_address, amount.to_string()],
            },
            "signature": {
                "type": "ed25519_signature",
                "public_key": "0x...",
                "signature": "0x...",
            },
        });

        let result = http_post_json(&url, body)?;
        result
            .get("hash")
            .or_else(|| result.get("transaction").and_then(|t| t.get("hash")))
            .and_then(|h| h.as_str())
            .map(|s| s.to_string())
            .ok_or_else(|| BridgeError::TransactionFailed("Failed to get transaction hash".to_string()))
    }

    pub fn is_valid_address(&self, address: &str) -> bool {
        // Aptos addresses are 0x followed by 64 hex characters
        address.starts_with("0x") && address.len() == 66 && address[2..].chars().all(|c| c.is_ascii_hexdigit())
    }
}
